src/section/shared_state.rs

```diff
@@ -0,0 +1,854 @@
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{
+    AgeCounter, EldersInfo, MemberState, SectionKeyInfo, SectionMap, SectionMembers,
+    SectionProofBlock, SectionProofChain, SectionProofSlice,
+};
+use crate::{
+    consensus::AccumulatingEvent,
+    error::Result,
+    id::{P2pNode, PublicId},
+    location::DstLocation,
+    network_params::NetworkParams,
+    relocation::{self, RelocateDetails},
+    xor_space::{Prefix, XorName, Xorable},
+};
+use std::{
+    collections::{BTreeMap, BTreeSet, VecDeque},
+    fmt::Debug,
+    net::SocketAddr,
+};
+
+/// Section state that is shared among all elders of a section via Parsec consensus.
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct SharedState {
+    /// Indicate whether nodes are shared state because genesis event was seen
+    #[serde(skip)]
+    pub handled_genesis_event: bool,
+    /// Our section's key history for Secure Message Delivery
+    pub our_history: SectionProofChain,
+    /// Info about all members of our section.
+    pub our_members: SectionMembers,
+    /// Info about known sections in the network.
+    pub sections: SectionMap,
+    /// Backlog of completed events that need to be processed when churn completes.
+    pub churn_event_backlog: VecDeque<AccumulatingEvent>,
+    /// Queue of pending relocations.
+    pub relocate_queue: VecDeque<RelocateDetails>,
+}
+
+impl SharedState {
+    pub fn new(
+        elders_info: EldersInfo,
+        bls_keys: bls::PublicKeySet,
+        ages: BTreeMap<PublicId, AgeCounter>,
+    ) -> Self {
+        let pk_info = SectionKeyInfo::from_elders_info(&elders_info, bls_keys.public_key());
+        let our_history = SectionProofChain::from_genesis(pk_info);
+        let our_key_info = our_history.last_key_info().clone();
+        let our_members = SectionMembers::new(&elders_info, &ages);
+
+        Self {
+            handled_genesis_event: false,
+            our_history,
+            sections: SectionMap::new(elders_info, our_key_info),
+            our_members,
+            churn_event_backlog: Default::default(),
+            relocate_queue: VecDeque::new(),
+        }
+    }
+
+    pub fn update(&mut self, new: Option<Self>) {
+        if self.handled_genesis_event {
+            log_or_panic!(
+                log::Level::Error,
+                "shared state update - genesis event already handled",
+            );
+        }
+
+        if let Some(new) = new {
+            if self.sections.has_our_history() && *self != new {
+                log_or_panic!(
+                    log::Level::Error,
+                    "shared state update - mismatch: old: {:?} --- new: {:?}",
+                    self,
+                    new
+                );
+            }
+
+            *self = new;
+        }
+
+        self.handled_genesis_event = true;
```

Comment:
Yes. A `None` input is sent to non-elders only because they don't need the full thing, but they still need the `handled_genesis_event` in case they are about to be promoted.